### PR TITLE
Add custom parameters benchmark page

### PR DESCRIPTION
**Pull request by Naarcha-AWS**
### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by gkamat — COMMENTED** [review]
The sections for Parameter Sources and Custom Runners should be completely separate, since they pertain to entirely different items.  Also, the examples should be re-written, to make them easier to understand.

We can provide new examples, but for now, the existing text should suffice.  Separating the sections can perhaps be done as part of the current PR.

**Comment by Naarcha-AWS** [comment]
@gkamat: I separated both files. Let me know what additional modifications you would like me to make.

File: _benchmark/user-guide/optimizing-benchmarks/custom-runners.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+layout: default
+title: Custom runners
+nav_order: 130
+parent: Optimizing benchmarks
+grand_parent: User guide
+---
+
+# Adding custom runners
+
+**Runners** are operation types performed on an OpenSearch cluster.
```
Comment (Naarcha-AWS) [inline _benchmark/user-guide/optimizing-benchmarks/custom-runners.md:11]: For adding support for an operation that doesn't exist in OSB already.

```suggestion
**Runners** are operation types performed on an OpenSearch cluster. 
```

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+layout: default
+title: Custom parameter sources 
+nav_order: 120
+parent: Optimizing benchmarks
+grand_parent: User guide
+---
+
+# Adding custom parameter sources 
+
+A **parameter source** is a source outside of OpenSearch Benchmark that provides parameters to an operation.
```
Comment (Naarcha-AWS) [inline _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:11]:
```suggestion
A **parameter source** is a source outside of OpenSearch Benchmark that provides parameters to an operation. You can add parameters to an existing operation in OpenSearch Benchmark or add custom parameters to a [runner]({{site.url}}{{site.baseurl}}/benchmark/user-guide/optimizing-benchmarks/custom-runners/).
```

**Review by gkamat — APPROVED** [review]
Will provide better examples of both items shortly.

File: _benchmark/user-guide/optimizing-benchmarks/custom-runners.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+layout: default
+title: Custom runners
+nav_order: 130
+parent: Optimizing benchmarks
+grand_parent: User guide
+---
+
+# Adding custom runners
+
+**Runners** are operation types performed on an OpenSearch cluster. You can add customer runners for operations that don't already exist in a workload.
```
Comment (IanHoang) [inline _benchmark/user-guide/optimizing-benchmarks/custom-runners.md:11]: > for operation types that do not exist in OpenSearch Benchmark (OSB).
Could we also link the types of operations that exist in OSB from the documentation?

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    ```
+
+3. In `workload.py` add the method or function for your custom parameter sources. The runner expects the parameter sources provided in step 2. The following example makes the workload use the `random_profession()` function for each operation that uses the parameter source `my-custom-term-param-source`. The function requires a `workload`, `params`, and `**movies`, a representation of the workload, as shown in the following example:
+
+
+    ```py
+    # In workload.py
+    import random
+
+    def random_profession(workload, params, **movies):
```
Comment (IanHoang) [inline _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:38]: random_profession -> random_pixar_movie

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+A **parameter source** is a source outside of OpenSearch Benchmark that provides parameters to an operation. 
+
+The `register_param_source` setting provides custom parameters for an operation. To use the `register_param_source` setting, use the following steps to modify `operations/default.json` and `workloads.py`.
+
+Adding custom parameters sources and runners modifies performance critical paths in OpenSearch Benchmark and therefore can lead to performance bottlenecks during texting. Carefully consider these changes before making them.
+{: .warning}
+
+
+1. Make sure that operation you want to modify exists in both `test_procedures/default.json` and `operations/default.json`.
+2. In both fields, add a `param-source` field to the operation, specifying the name of the parameter source. Add any additional fields which contain your custom parameters. For example, if you want to modify a term query operation that searches for a profession like “physician” to support a variety of professions supply a parameter source field called “custom-term-professions”. Then, add the additional professions as a list of strings ["mechanic", "physician", "nurse"], as shown in the following example:
```
Comment (IanHoang) [inline _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:20]: professions -> movies

Comment (IanHoang) [reply _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:20]: ["Inside Out", "The Incredibles", "Ratatouille"]

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+    ```json
+    {
+        "name": "term",
+        "operation-type": "search",
+        "param-source": "my-custom-term-param-source"
+        "professions": ["mechanic", "physician", "nurse"]
```
Comment (IanHoang) [inline _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:27]: "movies": ["Inside Out", "The Incredibles", "Ratatouille"]

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    }
+    ```
+
+3. In `workload.py` add the method or function for your custom parameter sources. The runner expects the parameter sources provided in step 2. The following example makes the workload use the `random_profession()` function for each operation that uses the parameter source `my-custom-term-param-source`. The function requires a `workload`, `params`, and `**movies`, a representation of the workload, as shown in the following example:
```
Comment (IanHoang) [inline _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:31]: random_profession() -> random_pixar_movie()

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                    }
+                }
+            },
+            "index": index_name,
+            "type": type_name,
+            "cache": params.get("cache", False)
+        }
+
+    def register(registry):
+        registry.register_param_source("my-custom-term-param-source", random_movie)
```
Comment (IanHoang) [inline _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:69]: random_pixar_movie

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+
+1. Make sure that operation you want to modify exists in both `test_procedures/default.json` and `operations/default.json`.
+2. In both fields, add a `param-source` field to the operation, specifying the name of the parameter source. Add any additional fields which contain your custom parameters. For example, if you want to modify a term query operation that searches for a profession like “physician” to support a variety of professions supply a parameter source field called “custom-term-professions”. Then, add the additional professions as a list of strings ["mechanic", "physician", "nurse"], as shown in the following example:
+
+    ```json
+    {
+        "name": "term",
+        "operation-type": "search",
+        "param-source": "my-custom-term-param-source"
```
Comment (IanHoang) [inline _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:26]: my-custom-term-param-source be renamed to custom-pixar-movies

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md (hunk 7 above)
Comment (IanHoang) [reply _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:31]: my-custom-term-param-source be renamed to custom-pixar-movies

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md (hunk 8 above)
Comment (IanHoang) [reply _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:69]: my-custom-term-param-source be renamed to custom-pixar-movies

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                    }
+                },
+                "index": self._index_name,
+                "type": self._type_name,
+                "cache": self._cache
+            }
+
+
+        def register(registry):
+            registry.register_param_source("my-custom-term-param-source", TermParamSource)
```
Comment (IanHoang) [inline _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:123]: my-custom-term-param-source be renamed to custom-pixar-movies

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+        index_name = params.get("index", default_index)
+        type_name = params.get("type", default_type)
+
+        # Provide all of the parameters the runner expects.
+        return {
+            "body": {
+                "query": {
+                    "term": {
+                        "body": "%s" % random.choice(params["movies"])
```
Comment (IanHoang) [inline _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:59]: movies -> kwargs

File: _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+4. (Optional) If you want more control over the function, you can add an additional class to `workload.py`, as shown in the following example:
+
+    ```py
+    # In workload.py
+    import random
+
+
+    class TermParamSource:
+        def __init__(self, workload, params, **movies):
```
Comment (IanHoang) [inline _benchmark/user-guide/optimizing-benchmarks/custom-parameter-sources.md:80]: movies --> kwargs

**Review by IanHoang — CHANGES_REQUESTED** [review]
Let's test this out before we merge this, as there might be some bugs.

**Pull request closed without merging**